lib/platform/linux-i2c.c

```diff
@@ -228,41 +244,134 @@ static uint8_t i2c_gas_cap_get(struct switchtec_dev *dev)
 		return (rx_buf[0] & GAS_TWI_MODE_MASK);
 }
 
-/* One I2C transaction can write a maximum of 28 bytes */
-#define I2C_MAX_WRITE 28
+/* One I2C transaction can write a maximum of 26 bytes */
+#define I2C_MAX_WRITE 24
 
-static void i2c_gas_write(struct switchtec_dev *dev, void __gas *dest,
-			  const void *src, size_t n)
+static uint8_t i2c_gas_data_write(struct switchtec_dev *dev, void __gas *dest,
+				  const void *src, size_t n, uint8_t tag)
 {
 	int ret;
 	struct switchtec_i2c *idev = to_switchtec_i2c(dev);
 
+	struct i2c_msg msg;
+	struct i2c_rdwr_ioctl_data wdata = {
+		.msgs = &msg,
+		.nmsgs = 1,
+	};
+
 	struct {
-		uint32_t gas_addr;
+		uint8_t command_code;
+		uint8_t byte_count;
+		uint8_t tag;
+		uint8_t offset[4];
 		uint8_t data[];
-	} *msg;
-
-	msg = malloc(sizeof(*msg) + n);
+	} *i2c_data;
 
-	msg->gas_addr = (uint32_t)(dest - (void __gas *)dev->gas_map);
+	uint32_t gas_addr = (uint32_t)(dest - (void __gas *)dev->gas_map);
 	assert(n <= I2C_MAX_WRITE);
-	assert((msg->gas_addr & 0x3) == 0);
 
-	memcpy(&msg->data, src, n);
-	msg->gas_addr = htobe32(msg->gas_addr);
+	/* PEC is the last byte */
+	i2c_data = malloc(sizeof(*i2c_data) + n + PEC_BYTE_COUNT);
 
-	ret = write(idev->fd, msg, sizeof(*msg) + n);
-	if (ret != sizeof(*msg) + n) {
+	i2c_data->command_code = CMD_GAS_WRITE;
+	i2c_data->byte_count = sizeof(i2c_data->tag) + sizeof(i2c_data->offset) + n;
+	i2c_data->tag = tag;
+
+	gas_addr = htobe32(gas_addr);
+	memcpy(i2c_data->offset, (uint8_t *)&gas_addr, sizeof(i2c_data->offset));
+	memcpy(&i2c_data->data, src, n);
+	msg.addr = idev->i2c_addr;
+	msg.flags = 0;
+	msg.len = sizeof(*i2c_data) + n + PEC_BYTE_COUNT;
+	msg.buf = (uint8_t *)i2c_data;
+
+	i2c_data->data[n] = i2c_msg_pec(&msg, msg.len - PEC_BYTE_COUNT, 0, true);
+
+	ret = ioctl(idev->fd, I2C_RDWR, &wdata);
+	if (ret < 0) {
 		/*
 		 * For now, this is all we can do. If it's a problem we'll
 		 * have to refactor all GAS access functions to return an
 		 * error code.
 		 */
-		perror("Could not communicate with I2C device");
-		exit(1);
+		return ret;
 	}
 
-	free(msg);
+	free(i2c_data);
+	return 0;
+}
+
+static uint8_t i2c_gas_write_status_get(struct switchtec_dev *dev, uint8_t tag)
+{
+	int ret;
+	struct switchtec_i2c *idev = to_switchtec_i2c(dev);
+	struct i2c_msg msgs[2];
+	struct i2c_rdwr_ioctl_data rwdata = {
+		.msgs = msgs,
+		.nmsgs = 2,
+	};
+
+	uint8_t command_code = CMD_GET_WRITE_STATUS;
+	uint8_t rx_buf[3];
+
+	uint8_t msg_0_pec, pec;
+	uint8_t retry_count = 0;
+
+	msgs[0].addr = msgs[1].addr = idev->i2c_addr;
+	msgs[0].flags = 0;
+	msgs[0].len = 1;
+	msgs[0].buf = &command_code;
+
+	msgs[1].flags = I2C_M_RD;
+	msgs[1].len = 3;
+	msgs[1].buf = rx_buf;
+
+	do {
+		usleep(2000);
+		ret = ioctl(idev->fd, I2C_RDWR, &rwdata);
+		if (ret < 0) {
+			/*
+			 * For now, this is all we can do. If it's a problem we'll
+			 * have to refactor all GAS access functions to return an
+			 * error code.
+			 */
```

Comment:
This comment doesn't seem accurate... The calling function does it in a loop. The calling function really needs a comment.